File: src/leetcode/trees/c/226.c
#include "eval.c"
#include "stddef.h"

struct Node *invertTree(struct Node *root) {
  if (root != NULL) {
    // Swap Left and Right subTree
    struct Node *tmp = root->left;
    root->left = root->right;
    root->right = tmp;
    // Recursively invert Left and Right subTree
    invertTree(root->left);
    invertTree(root->right);
  }

  return root;
}

int main(int argc, char *argv[]) {
  // Parse arguments
  int output_size = 0, input_size = 0, expected_output_size = 0;
  int *input = parse_single_array(argv[1], &input_size);
  int *expected_output = parse_single_array(argv[2], &expected_output_size);

  // Create Trees
  Node *inputTree = constructor(input, input_size);
  Node *invertedTree = invertTree(inputTree);

  if (invertedTree == 0) {
    int *output = (int *)malloc(0);
    return 0;
  }

  // Convert Node* to int*
  int *output = destructor(invertedTree, &output_size);
  if (arrcmp(output, expected_output, output_size, expected_output_size) != 0) {
    return 1;
  }

  return 0;
}

File: src/leetcode/trees/rust/src/_226.rs
use crate::eval::{constructor, destructor, Tree, TreeNode, TreeTreeTestCase};
use crate::utils::parse_json_array;

trait TreeNodeSwap {
    fn swap(&mut self);
    fn swap_all(&mut self);
}

impl TreeNodeSwap for TreeNode {
    fn swap(&mut self) {
        std::mem::swap(&mut self.left, &mut self.right)
    }

    fn swap_all(&mut self) {
        self.left.as_mut().map(|node| node.borrow_mut().swap_all());
        self.right.as_mut().map(|node| node.borrow_mut().swap_all());
        self.swap();
    }
}

fn invert_tree(root: Option<Tree>) -> Option<Tree> {
    root.map(|node| {
        node.borrow_mut().swap_all();
        node
    })
}

pub fn main(test_cases: Vec<TreeTreeTestCase>) {
    for (_, test_case) in test_cases.iter().enumerate() {
        let input = parse_json_array::<i64>(&test_case.input);
        let expected_output = parse_json_array::<i64>(&test_case.output);

        let input_tree = constructor(&input);
        let inverted_tree = invert_tree(input_tree);
        let output = destructor(inverted_tree);

        assert_eq!(output, expected_output, "Test Case {:?} failed", input);
    }
    println!("Tests ran successfully for problem ID 226");
}
